src/qt3utils/datagenerators/princeton.py

```diff
@@ -0,0 +1,344 @@
+    #      - stop the scan 
+        
+    def save_current_experiment(self) -> None:
+        self.experiment.Save()
+    
+    def save_new_experient(self, value) -> None:
+        self.experiment.SaveAs(value)
+
+    def stop_scan(self) -> None:
+        self.experiment.Stop()
+
+    #NOTE: Thw while loop in here will not be needed once you fix the "FileNameGeneration" problem.
+    def start_acquisition_and_wait(self) -> None:
+        """
+        Starts the acquisition process and waits until it is completed before carrying on.
+        """
+        acquisition_time = self.get(lf.AddIns.CameraSettings.ShutterTimingExposureTime)
+        num_frames = self.get(lf.AddIns.ExperimentSettings.AcquisitionFramesToStore)
+        self.experiment.Acquire()
+        
+        sleep(0.001 * acquisition_time * num_frames)  # waiting for the exposure duration
+        
+        while self.experiment.IsRunning:
+            sleep(0.1)  # checking if the experiment is still running
+
+    def process_acquired_data(self) -> Union[np.ndarray, Any]:
+        """
+        Processes the most recently acquired data and returns it as a numpy array.
+        """
+        last_file = self._application.FileManager.GetRecentlyAcquiredFileNames()[0]
+        curr_image = self._application.FileManager.OpenFile(last_file, FileAccess.Read)
+
+        if curr_image.Regions.Length == 1:
+            if curr_image.Frames == 1:
+                return self._process_single_frame(curr_image)
+            else:
+                return self._process_multiple_frames(curr_image)
+        else:
+            raise Exception('curr_image.Regions is not valid. Please retry.')
+
+    def _process_single_frame(self, curr_image: Any) -> np.ndarray:
+        """
+            **Single Frame**:
+            - Data is returned as a 2D array representing raw counts from each pixel.
+            - A vertical section (spanning 400 pixels) represents a range for wavelength averaging.
+        """
+        frame = curr_image.GetFrame(0, 0)
+        return np.reshape(np.fromiter(frame.GetData(), dtype='uint16'), [frame.Width, frame.Height], order='F')
+
+    def _process_multiple_frames(self, curr_image: Any) -> np.ndarray:
+        """
+            **Multiple Frames**:
+            - Data from successive exposures is added as a new dimension, then returns a 3D array.
+            - Averaging across frames can be done by summing over this additional dimension.
+        """
+        data = np.array([])
+        for i in range(curr_image.Frames):
+            frame = curr_image.GetFrame(0, i)
+            new_frame = np.reshape(np.fromiter(frame.GetData(), dtype='uint16'), [frame.Width, frame.Height], order='F')
+            data = np.dstack((data, new_frame)) if data.size else new_frame
+        return data
+
+    #NOTE: May not need to call the 'start_acquisition_and_wait' method if you fix the 'FileNameGeneration' issue
+    def acquire(self) -> Union[np.array, None]:
```

Comment:
In what scenario will this function return `None`?

Reply:
Apologies, I was going to fix this around the time you told me you were reviewing so I left it